### PR TITLE
Add defaults to pos-only parameters in the stub, even if the parameter name is incorrect in the stub

### PR DESCRIPTION
**Pull request by AlexWaygood**
If the number of parameters in the stub matches the number of parameters at runtime, and there are no variadic parameters at runtime, it should be safe to add defaults to the stub *even if* the stub has an incorrect name for the parameter

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Shouldn't we fix the parameter name in the stub instead in this case?

**Comment by AlexWaygood** [comment]
> Shouldn't we fix the parameter name in the stub instead in this case?

We should do that as well, yes.

The underlying issue here is that stubtest doesn't precisely verify parameter names for positional-only parameters. But that doesn't feel like the responsibility of stubdefaulter?

I don't feel too strongly though, we can close this if you like :) The patch was pretty good at exposing that we for some reason have quite a few `__get__` methods with incorrect parameter names in typeshed, so it's been useful in any case!

**Comment by AlexWaygood** [comment]
We could possibly have this patch autofix the parameter names as well...

**Comment by JelleZijlstra** [comment]
I guess stubtest works that way because pos-only parameter names are often inconsistent across versions (e.g., when AC is applied to a function) or across the C and Python versions of a function. For those cases, it's reasonable for stubdefaulter to also pick up the default.

Could you limit this patch to only positional-only parameters? That feels safer.

File: stubdefaulter.py [hunk 1]
```diff
@@ -61,19 +61,61 @@ def infer_value_of_node(node: libcst.BaseExpression) -> object:
...
+            chain(
+                self.stub_params.posonly_params,
+                self.stub_params.params,
+                self.stub_params.kwonly_params,
+            )
+        )
+        num_stub_params = len(all_stub_params)
+        num_runtime_params = len(all_runtime_parameters)
+        if num_stub_params == num_runtime_params:
+            n = next(i for i, param in enumerate(all_stub_params) if param == node)
```
Comment (JelleZijlstra) [inline stubdefaulter.py:110]: `all_stub_param.index(node)`?

File: stubdefaulter.py [hunk 2]
```diff
@@ -61,19 +61,61 @@ def infer_value_of_node(node: libcst.BaseExpression) -> object:
...
+            return next(islice(all_runtime_parameters, n, n + 1))
```
Comment (JelleZijlstra) [inline stubdefaulter.py:111]: I would just make `all_runtime_parameters` a list and use `[]`.

File: stubdefaulter.py [hunk 3]
```diff
@@ -61,19 +61,61 @@ def infer_value_of_node(node: libcst.BaseExpression) -> object:
...
-        param: inspect.Parameter | None = None
         try:
-            param = self.sig.parameters[param_name]
+            return self.sig.parameters[param_name]
         except KeyError:
-            if param_name.startswith("__") and not param_name.endswith("__"):
-                param = self.sig.parameters.get(param_name[2:])
+            pass
+
+        if (
```
Comment (JelleZijlstra) [inline stubdefaulter.py:77]: This code could use some comments on what kinds of heuristics we're using to match up parameters and why we think that's safe.

File: test_stubdefaulter.py [hunk 4]
```diff
@@ -74,6 +74,9 @@ def useless_runtime(*args, **kwargs):
             )
     if 'enum_default' in kwargs and kwargs['enum_default'] is not re.ASCII:
         raise ValueError("FOOL")
+
+def incorrect_parameter_names_in_stub(x="foo", y="bar", *, z="baz"):
```
Comment (JelleZijlstra) [inline test_stubdefaulter.py:78]: I was thinking we'd do this only if the parameter is pos-only both in the stub and at runtime. Does that make the heuristic significantly less useful? I know we sometimes mark things as pos-only in the stub even if they technically aren't at runtime.

Comment (AlexWaygood) [reply test_stubdefaulter.py:78]: I don't think that makes the heuristic significantly less useful; I think it makes sense, and I'm happy to make the change.

Practically speaking, I'm not sure how I'd write a test for it that passes on Python 3.7 in the context of this test suite, though!

Comment (JelleZijlstra) [reply test_stubdefaulter.py:78]: `if sys.version_info >= (3. 8):`?

Or compile a C extension that makes a real positional-only argument :D

File: stubdefaulter.py (hunk 3 above)
Comment (AlexWaygood) [reply stubdefaulter.py:86]: Added some comments, though they may be too verbose

File: test_stubdefaulter.py (hunk 4 above)
Comment (AlexWaygood) [reply test_stubdefaulter.py:78]: I'll have a play around tomorrow; it's late here now :)

Comment (JelleZijlstra) [reply test_stubdefaulter.py:78]: I think this works actually:

```
>>> def f(x): pass
... 
>>> f.__signature__ = inspect.Signature([inspect.Parameter("x", inspect.Parameter.POSITIONAL_ONLY)])
>>> 
>>> inspect.signature(f)
<Signature (x, /)>
```

File: stubdefaulter.py (hunk 3 above)
Comment (JelleZijlstra) [reply stubdefaulter.py:86]: Thanks, I like the comments.

File: test_stubdefaulter.py (hunk 4 above)
Comment (AlexWaygood) [reply test_stubdefaulter.py:78]: Oh wow, I had no idea:

```pycon
(py37) C:\Users\alexw\coding>python
Python 3.7.16 (default, Jan 17 2023, 16:06:28) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo(x): ...
...
>>> import inspect
>>> foo.__signature__ = inspect.Signature([inspect.Parameter("x", inspect.Parameter.POSITIONAL_ONLY)])
>>> inspect.signature(foo)
<Signature (x, /)>
>>> def bar(x, /): ...
  File "<stdin>", line 1
    def bar(x, /): ...
               ^
SyntaxError: invalid syntax
```

??

File: stubdefaulter.py [hunk 5]
```diff
@@ -61,19 +61,87 @@ def infer_value_of_node(node: libcst.BaseExpression) -> object:
...
+        # that have the same name as the stub parameter,
+        # or that have the same name with `__` prepended.
+        # Fall back to the nth parameter of the runtime signature
+        # (where n is the index of the stub parameter in the stub signature),
+        # iff all the following conditions are true:
+        #
+        # - The number of parameters at runtime == the number of parameters in the stub
+        # - There are no variadic parameters (*args or **kwargs) at runtime or in the stub
+        # - The parameter is marked as being pos-only in the stub,
+        #   either through PEP 570 syntax or through a parameter name starting with `__`
```
Comment (AlexWaygood) [inline stubdefaulter.py:108]:
```suggestion
        #   either through PEP 570 syntax or through a parameter name starting with `__`
        # - The parameter is also positional-only at runtime
```

**Comment by AlexWaygood** [comment]
If I run this patch on https://github.com/python/typeshed/commit/74112dc18919084093d7aed3f26456ccf4598cf8, it produces this diff:

<details>

```diff
diff --git a/stdlib/_decimal.pyi b/stdlib/_decimal.pyi
index b8208fe18..dbd29f935 100644
--- a/stdlib/_decimal.pyi
+++ b/stdlib/_decimal.pyi
@@ -89,7 +89,7 @@ class Decimal:
     def __mul__(self, __other: _Decimal) -> Decimal: ...
     def __neg__(self) -> Decimal: ...
     def __pos__(self) -> Decimal: ...
-    def __pow__(self, __other: _Decimal, __modulo: _Decimal | None = ...) -> Decimal: ...
+    def __pow__(self, __other: _Decimal, __modulo: _Decimal | None = None) -> Decimal: ...
     def __radd__(self, __other: _Decimal) -> Decimal: ...
     def __rdivmod__(self, __other: _Decimal) -> tuple[Decimal, Decimal]: ...
     def __rfloordiv__(self, __other: _Decimal) -> Decimal: ...
@@ -116,7 +116,7 @@ class Decimal:
     def __floor__(self) -> int: ...
     def __ceil__(self) -> int: ...
     def fma(self, other: _Decimal, third: _Decimal, context: Context | None = None) -> Decimal: ...
-    def __rpow__(self, __other: _Decimal, __context: Context | None = ...) -> Decimal: ...
+    def __rpow__(self, __other: _Decimal, __context: Context | None = None) -> Decimal: ...
     def normalize(self, context: Context | None = None) -> Decimal: ...
     def quantize(self, exp: _Decimal, rounding: str | None = None, context: Context | None = None) -> Decimal: ...
     def same_quantum(self, other: _Decimal, context: Context | None = None) -> bool: ...
diff --git a/stdlib/builtins.pyi b/stdlib/builtins.pyi
index a8bedc837..e340cbe62 100644
--- a/stdlib/builtins.pyi
+++ b/stdlib/builtins.pyi
@@ -117,7 +117,7 @@ class staticmethod(Generic[_R_co]):
     @property
     def __isabstractmethod__(self) -> bool: ...
     def __init__(self: staticmethod[_R_co], __f: Callable[..., _R_co]) -> None: ...
-    def __get__(self, __obj: _T, __type: type[_T] | None = ...) -> Callable[..., _R_co]: ...
+    def __get__(self, __obj: _T, __type: type[_T] | None = None) -> Callable[..., _R_co]: ...
     if sys.version_info >= (3, 10):
         __name__: str
         __qualname__: str
@@ -131,7 +131,7 @@ class classmethod(Generic[_R_co]):
     @property
     def __isabstractmethod__(self) -> bool: ...
     def __init__(self: classmethod[_R_co], __f: Callable[..., _R_co]) -> None: ...
-    def __get__(self, __obj: _T, __type: type[_T] | None = ...) -> Callable[..., _R_co]: ...
+    def __get__(self, __obj: _T, __type: type[_T] | None = None) -> Callable[..., _R_co]: ...
     if sys.version_info >= (3, 10):
         __name__: str
         __qualname__: str
@@ -1185,7 +1185,7 @@ class property:
     def getter(self, __fget: Callable[[Any], Any]) -> property: ...
     def setter(self, __fset: Callable[[Any, Any], None]) -> property: ...
     def deleter(self, __fdel: Callable[[Any], None]) -> property: ...
-    def __get__(self, __obj: Any, __type: type | None = ...) -> Any: ...
+    def __get__(self, __obj: Any, __type: type | None = None) -> Any: ...
     def __set__(self, __obj: Any, __value: Any) -> None: ...
     def __delete__(self, __obj: Any) -> None: ...

diff --git a/stdlib/types.pyi b/stdlib/types.pyi
index d529b3d9a..526a4845e 100644
--- a/stdlib/types.pyi
+++ b/stdlib/types.pyi
@@ -454,7 +454,7 @@ class WrapperDescriptorType:
     @property
     def __objclass__(self) -> type: ...
     def __call__(self, *args: Any, **kwargs: Any) -> Any: ...
-    def __get__(self, __obj: Any, __type: type = ...) -> Any: ...
+    def __get__(self, __obj: Any, __type: type = None) -> Any: ...

 @final
 class MethodWrapperType:
@@ -536,7 +536,7 @@ class GetSetDescriptorType:
     def __qualname__(self) -> str: ...
     @property
     def __objclass__(self) -> type: ...
-    def __get__(self, __obj: Any, __type: type = ...) -> Any: ...
+    def __get__(self, __obj: Any, __type: type = None) -> Any: ...
     def __set__(self, __instance: Any, __value: Any) -> None: ...
     def __delete__(self, __obj: Any) -> None: ...

@@ -548,7 +548,7 @@ class MemberDescriptorType:
     def __qualname__(self) -> str: ...
     @property
     def __objclass__(self) -> type: ...
-    def __get__(self, __obj: Any, __type: type = ...) -> Any: ...
+    def __get__(self, __obj: Any, __type: type = None) -> Any: ...
     def __set__(self, __instance: Any, __value: Any) -> None: ...
     def __delete__(self, __obj: Any) -> None: ...
```

</details>

**Review by JelleZijlstra — APPROVED** [review]
Thanks!

**Pull request merged**